elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/Schema.java

```diff
@@ -0,0 +1,290 @@
+     *
+     * @return {@code true} if the field is a simple metric
+     */
+    public boolean isBaseMetric(String fieldName) {
+        if (!isMetricField(fieldName)) {
+            return false;
+        }
+
+        return !getEntityDictionary().attributeOrRelationAnnotationExists(
+                getEntityClass(),
+                fieldName,
+                MetricComputation.class
+        );
+    }
+
+    /**
+     * Constructs a new {@link Metric} instance.
+     *
+     * @param metricField  The entity field of the metric being constructed
+     * @param cls  The entity that contains the metric being constructed
+     * @param entityDictionary  The auxiliary object that offers binding info used to construct this
+     * {@link Metric}
+     *
+     * @return a {@link Metric}
+     */
+    protected Metric constructMetric(String metricField, Class<?> cls, EntityDictionary entityDictionary) {
+        Meta metaData = entityDictionary.getAttributeOrRelationAnnotation(cls, Meta.class, metricField);
+        Class<?> fieldType = entityDictionary.getType(cls, metricField);
+
+        // get all metric aggregations
+        List<Class<? extends Aggregation>> aggregations = Arrays.stream(
+                entityDictionary.getAttributeOrRelationAnnotation(
+                        cls,
+                        MetricAggregation.class,
+                        metricField
+                ).aggregations()
+        )
+                .collect(
+                        Collectors.collectingAndThen(
+                                Collectors.toList(),
+                                Collections::unmodifiableList
+                        )
+                );
+
+        // make sure we have at least 1 aggregation so we can generate static SQL functions
+        if (aggregations.isEmpty()) {
+            String message = String.format("'%s' has no aggregation.", metricField);
+            log.error(message);
+            throw new IllegalStateException(message);
+        }
+
+        return new BaseMetric(
+                metricField,
+                metaData,
+                fieldType,
+                aggregations,
+                BaseMetric.functionFormat(aggregations.get(0))
+        );
+    }
+
+    /**
+     * Constructs and returns a new instance of {@link Dimension}.
+     *
+     * @param dimensionField  The entity field of the dimension being constructed
+     * @param cls  The entity that contains the dimension being constructed
+     * @param entityDictionary  The auxiliary object that offers binding info used to construct this
+     * {@link Dimension}
+     *
+     * @return a {@link Dimension}
+     */
+    protected Dimension constructDimension(String dimensionField, Class<?> cls, EntityDictionary entityDictionary) {
+        Meta metaData = entityDictionary.getAttributeOrRelationAnnotation(cls, Meta.class, dimensionField);
+        Class<?> fieldType = entityDictionary.getType(cls, dimensionField);
```

Comment:
There are two getType methods in the dictionary: getType and getParamterizedType.  This might not work for `Set<Player>` for example.  We should test this.

Reply:
I think this still might be a bug.  Does it work?

Reply:
Oh sorry I keep forgetting this. Let me add a test to it in the next commit